WebDriver BiDi: support for browsingContext lifecycle domContentLoaded & load

AnujPanta1 opened the pull request:
#### 75b2f2cc0c78d2400cafa6defcfb1f610d6b299e
<pre>
WebDriver BiDi: support for browsingContext lifecycle domContentLoaded &amp; load
<a href="https://bugs.webkit.org/show_bug.cgi?id=297281">https://bugs.webkit.org/show_bug.cgi?id=297281</a>
<a href="https://rdar.apple.com/158155686">rdar://158155686</a>

Reviewed by BJ Burg.

WebDriver BiDi events hooked into WebKit&apos;s page load lifecycle callbacks:
documentLoadedForFrame() now triggers domContentLoaded event,
didFinishLoadForFrame() now triggers load event.

This ensures we&apos;re correctly firing these events while not interfering
with existing functionality. We did add NavigationIdenfitier to some WebAutomationSession
functions to carry that information from WebPageProxy.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::navigationIDToProtocolString):
(WebKit::WebAutomationSession::documentLoadedForFrame):
(WebKit::WebAutomationSession::loadCompletedForFrame):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFinishDocumentLoadForFrame):
(WebKit::WebPageProxy::didFinishLoadForFrame):

Canonical link: <a href="https://commits.webkit.org/299810@main">https://commits.webkit.org/299810@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2cb694a5f894479c5809c7ef99152fcdb8f803ad

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120101 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39795 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30446 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126445 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72174 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40491 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48372 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91210 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60517 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123053 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32343 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107686 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71761 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31379 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70073 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101822 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25978 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129354 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47021 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35669 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99829 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47387 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103873 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99672 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25337 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45123 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23148 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43652 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46883 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52589 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46351 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49698 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48035 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -910,6 +928,17 @@ void WebAutomationSession::documentLoadedForFrame(const WebFrameProxy& frame)
 
+void WebAutomationSession::loadCompletedForFrame(const WebFrameProxy& frame, std::optional<WebCore::NavigationIdentifier> navigationID, WallTime timestamp)
+{
+#if ENABLE(WEBDRIVER_BIDI)
```
I used the flags here because this is a classic function and we're adding Bibi logic

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -862,6 +862,16 @@ void WebAutomationSession::reloadBrowsingContext(const Inspector::Protocol::Auto
     waitForNavigationToCompleteOnPage(*page, pageLoadStrategy, pageLoadTimeout, WTFMove(callback));
 }
 
+static String convertNavigationIDToString(std::optional<WebCore::NavigationIdentifier> navigationID)
```
this got left over, already included in this PR (https://github.com/WebKit/WebKit/pull/49280/files#diff-8bd8c79ff233a5f127fcdeeb7bb3abc7dcb8d0ca2d2c828130197d10e3f53302), I will remove before reviewing, left in for testing bots

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -891,8 +901,16 @@ void WebAutomationSession::navigationOccurredForFrame(const WebFrameProxy& frame
+#if ENABLE(WEBDRIVER_BIDI)
+    String browsingContextHandle = handleForWebFrameProxy(frame);
+    String navigationIDString = convertNavigationIDToString(navigationID);
+    double timestampInMilliseconds = timestamp.secondsSinceEpoch().milliseconds();
```
Please inline these local variables into the call below.

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -910,6 +928,17 @@ void WebAutomationSession::documentLoadedForFrame(const WebFrameProxy& frame)
+    String navigationIDString = convertNavigationIDToString(navigationID);
+    double timestampInMilliseconds = timestamp.secondsSinceEpoch().milliseconds();
+    String url = frame.url().string();
+    m_bidiProcessor->browsingContextDomainNotifier().load(browsingContextHandle, navigationIDString, timestampInMilliseconds, url);
```
Ditto

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -176,6 +176,29 @@
+            "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/browsing_context/dom_content_loaded",
+            "parameters": [
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext" },
+                { "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "nullable": true },
```
Nit: I'm not sure if this is actually nullable in practice, let's drop the "null" if we aren't actually setting it null.

(My understanding from the previous PR review is that `null` value here is only acceptable for events in which the navigation has failed or cancelled without making progress.)

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -176,6 +176,29 @@
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext" },
+                { "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "nullable": true },
+                { "name": "timestamp", "type": "number", "description": "Time value in milliseconds representing the current date and time in UTC." },
+                { "name": "url", "type": "string" }
```
Might want to specify if this is  the provisional, final (committed), or "current" URL?

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -176,6 +176,29 @@
+            "spec": "https://w3c.github.io/webdriver-bidi/#event-browsingContext-load",
+            "parameters": [
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext" },
+                { "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "nullable": true },
```
Ditto earlier comments

burg commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7343,7 +7343,7 @@ void WebPageProxy::didFinishDocumentLoadForFrame(IPC::Connection& connection, Fr
     if (m_controlledByAutomation) {
         if (RefPtr automationSession = m_configuration->processPool().automationSession())
-            automationSession->documentLoadedForFrame(*frame);
+            automationSession->documentLoadedForFrame(*frame, navigationID, timestamp);
```
Please refactor (as you go) to use the new `activeAutomationSession()` helper method so that we can reduce the indentation by one level here.

burg commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7442,8 +7442,10 @@ void WebPageProxy::didFinishLoadForFrame(IPC::Connection& connection, FrameIdent
 
         if (m_controlledByAutomation) {
-            if (RefPtr automationSession = m_configuration->processPool().automationSession())
+            if (RefPtr automationSession = m_configuration->processPool().automationSession()) {
```
Ditto.

burg commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7442,8 +7442,10 @@ void WebPageProxy::didFinishLoadForFrame(IPC::Connection& connection, FrameIdent
-            if (RefPtr automationSession = m_configuration->processPool().automationSession())
+            if (RefPtr automationSession = m_configuration->processPool().automationSession()) {
                 automationSession->navigationOccurredForFrame(*frame);
+                automationSession->loadCompletedForFrame(*frame, navigationID, WallTime::now());
```
It seems silly to me that we basically do the same call twice here- can we instead make the call here `session->loadCompletedForFrame(...)` and call out to the Bidi agent if needed underneath that?

burg reviewed (approved):
Functionally this PR is looking good. Let's do another iteration to fix the EWS failures for !ENABLE_WEBDRIVER_BIDI and address the style nits. After making the improvements that we made other PR, I think this one will be really short and sweet.

lauromoura commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -896,8 +896,23 @@ void WebAutomationSession::navigationOccurredForFrame(const WebFrameProxy& frame
+void WebAutomationSession::documentLoadedForFrame(const WebFrameProxy& frame, std::optional<WebCore::NavigationIdentifier> navigationID, WallTime timestamp)
 {
+#if ENABLE(WEBDRIVER_BIDI)
+    m_bidiProcessor->browsingContextDomainNotifier().domContentLoaded(handleForWebFrameProxy(frame), navigationIDToProtocolString(navigationID), timestamp.secondsSinceEpoch().milliseconds(), frame.url().string());
```
```suggestion
    m_bidiProcessor->browsingContextDomainNotifier().domContentLoaded(handleForWebFrameProxy(frame), navigationIDToProtocolString(navigationID), std::trunc(timestamp.secondsSinceEpoch().milliseconds()), frame.url().string());
```

lauromoura commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -915,6 +930,13 @@ void WebAutomationSession::documentLoadedForFrame(const WebFrameProxy& frame)
+void WebAutomationSession::loadCompletedForFrame(const WebFrameProxy& frame, std::optional<WebCore::NavigationIdentifier> navigationID, WallTime timestamp)
+{
+#if ENABLE(WEBDRIVER_BIDI)
+    m_bidiProcessor->browsingContextDomainNotifier().load(handleForWebFrameProxy(frame), navigationIDToProtocolString(navigationID), timestamp.secondsSinceEpoch().milliseconds(), frame.url().string());
```
```suggestion
    m_bidiProcessor->browsingContextDomainNotifier().load(handleForWebFrameProxy(frame), navigationIDToProtocolString(navigationID), std::trunc(timestamp.secondsSinceEpoch().milliseconds()), frame.url().string());
```

lauromoura reviewed (commented):
Looks good, just a minor change needed in the timestamp, as the spec defines "time values" as a integral numbers in https://tc39.es/ecma262/#sec-time-values-and-time-range

> An ECMAScript time value [is a Number](https://tc39.es/ecma262/#sec-ecmascript-language-types-number-type), either a [finite](https://tc39.es/ecma262/#finite) [integral Number](https://tc39.es/ecma262/#integral-number) representing an instant in time to millisecond precision...

In fact, the WPT tests for navigation timestamps expect int values: https://github.com/web-platform-tests/wpt/blob/c7966f035094fb955c40c6e02185893c12ad8aac/webdriver/tests/bidi/browsing_context/__init__.py#L72

AnujPanta1 committed:
WebDriver BiDi: support for browsingContext lifecycle domContentLoaded & load
https://bugs.webkit.org/show_bug.cgi?id=297281
rdar://158155686

Reviewed by BJ Burg.

WebDriver BiDi events hooked into WebKit's page load lifecycle callbacks:
documentLoadedForFrame() now triggers domContentLoaded event,
didFinishLoadForFrame() now triggers load event.

This ensures we're correctly firing these events while not interfering
with existing functionality. We did add NavigationIdenfitier to some WebAutomationSession
functions to carry that information from WebPageProxy.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::navigationIDToProtocolString):
(WebKit::WebAutomationSession::documentLoadedForFrame):
(WebKit::WebAutomationSession::loadCompletedForFrame):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFinishDocumentLoadForFrame):
(WebKit::WebPageProxy::didFinishLoadForFrame):

Canonical link: https://commits.webkit.org/299810@main
